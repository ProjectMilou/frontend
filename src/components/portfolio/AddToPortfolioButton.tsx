import React from 'react';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import Checkbox from '@material-ui/core/Checkbox';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import MuiDialogTitle from '@material-ui/core/DialogTitle';
import MuiDialogContent from '@material-ui/core/DialogContent';
import MuiDialogActions from '@material-ui/core/DialogActions';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import { useTranslation } from 'react-i18next';
import * as API from '../../portfolio/APIClient';
import { errorMessageKey, errorTitleKey } from '../../Errors';
import ProgressButton from './ProgressButton';

// stylesheet for the entire add to portfolio component
const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    addButton: {
      padding: '0.25rem 1rem',
      backgroundColor: '#3fbcf2',
      '&:hover': {
        backgroundColor: '#84d4f7',
      },
      whiteSpace: 'nowrap',
    },
    Dialog: {
      minWidth: '40rem',
    },
    dialogHeader: {
      margin: 0,
      padding: theme.spacing(2),
      display: 'flex',
      justifyContent: 'center',
    },
    // dialog header
    headerCloseButton: {
      position: 'relative',
      color: theme.palette.grey[500],
    },
    span: {
      display: 'flex',
    },
    text: {
      margin: 'auto',
    },
    // dialog body
    DialogContent: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'space-between',
      margin: 0,
      padding: theme.spacing(1),
    },
    // List component
    List: {
      width: '100%',
      backgroundColor: 'white',
    },
    listItem: {
      width: '100%',
      margin: '0.75rem 0',
      height: '3.25rem',
    },
    // error helper text
    helperText: {
      position: 'absolute',
      top: '100%',
      right: '-50%',
      whiteSpace: 'nowrap',
    },
    dialogFooter: {
      margin: 0,
      padding: theme.spacing(2),
      display: 'flex',
      justifyContent: 'end',
    },
  })
);

type ListEntryProps = {
  id: string;
  name: string;
  checked: CheckedType;
  handleToggle: (id: string) => () => void;
  changeAmount: (id: string, amount: string) => void;
  pattern: RegExp;
};

const ListEntry: React.FC<ListEntryProps> = ({
  id,
  name,
  checked,
  handleToggle,
  changeAmount,
  pattern,
}) => {
  const classes = useStyles();

  return (
    <ListItem role={undefined} dense className={classes.listItem}>
      <ListItemIcon>
        <Checkbox
          edge="start"
          checked={typeof checked[id] !== 'undefined'}
          tabIndex={-1}
          disableRipple
          inputProps={{
            'aria-labelledby': `checkbox-list-label-${id}`,
          }}
          color="secondary"
          onClick={handleToggle(id)}
        />
      </ListItemIcon>
      <ListItemText primary={name} />
      <TextField
        variant="outlined"
        value={checked[id]}
        error={!pattern.test(checked[id])}
        helperText={
          pattern.test(checked[id]) ? '' : 'Please enter a positive number!'
        }
        FormHelperTextProps={{
          className: classes.helperText,
        }}
        onChange={(e) => {
          changeAmount(id, e.target.value);
        }}
        size="small"
        margin="dense"
        inputProps={{
          maxLength: 5,
          style: {
            textAlign: 'center',
            paddingRight: 0,
            paddingLeft: '0.2rem',
          },
        }}
        style={{
          display: typeof checked[id] !== 'undefined' ? 'block' : 'none',
          width: '4rem',
        }}
      />
    </ListItem>
  );
};

type CheckedType = {
  // id of portfolio: qty of stock;
  [id: string]: string;
};

type AddToPortfolioButtonType = {
  isin: string;
  token: string;
};

// returns the add to portfolio button and all subcomponents including the dialog window
const AddToPortfolioButton: React.FC<AddToPortfolioButtonType> = ({
  isin,
  token,
}) => {
  // style and translation hooks
  const classes = useStyles();
  const { t } = useTranslation();
  // pattern for testing input fields
  const pattern = /^(?:[1-9]\d*|0)?(?:\.\d+)?$/;

  // the list of portfolios belonging to the user
  const [portfolios, setPortfolios] = React.useState<API.PortfolioOverview[]>(
    []
  );

  // the state that controls whether the dialog window is open or closed
  const [open, setOpen] = React.useState(false);
  // true when waiting for an API response
  const [loading, setLoading] = React.useState<boolean>(false);
  const [error, setError] = React.useState<Error | undefined>();
  // keep track of which portfolios have been selected and the amount in the textfield
  const [checked, setChecked] = React.useState<CheckedType>({});

  // defines function to fetch from api
  const isMounted = React.useRef(true);
  const fetch = async () => {
    setError(undefined);
    try {
      const p = await API.list(token);
      if (isMounted.current) {
        setPortfolios(p);
      }
    } catch (e) {
      if (isMounted.current) {
        setError(e);
      }
    }
  };

  // opon mount fetch list of portfolios from api
  React.useEffect(() => {
    fetch();
    return () => {
      isMounted.current = false;
    };
    // deps must be empty because the function should only be called on mount.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const openDialog = () => {
    // reset the dialog's state on (re-)open
    setLoading(false);
    setError(undefined);
    setChecked({});
    setOpen(true);
  };

  // handle saving selections
  const handleSave = async () => {
    setLoading(true);

    try {
      // parse the checked state into the form expected by the api
      const modifications: API.PortfoliosQty[] = Object.entries(checked).map(
        ([id, qty]) => ({
          id,
          // TODO: use react-number-format for validation,
          //  store numbers instead of strings, avoid converting
          qty: Number.parseFloat(qty),
        })
      );
      // make api put call
      await API.stock(token, isin, modifications);
      setOpen(false);
    } catch (e) {
      // on error stop the loading animation and display error to user
      setLoading(false);
      setError(e);
    }
  };

  // handles checking and unchecking portfolios
  const handleCheckToggle = (id: string) => () => {
    const newChecked = { ...checked };

    if (typeof newChecked[id] === 'undefined') {
      // if unchecked, check it with an initial value of 1
      newChecked[id] = '1';
    } else {
      // if checked remove check
      delete newChecked[id];
    }

    setChecked(newChecked);
  };

  //  a function that sets the amount everytime a textfield is changed
  const changeAmount = (id: string, amount: string) => {
    // when a user changes the input field the checked state gets updated
    setChecked({ ...checked, [id]: amount });
  };

  return (
    <div>
      <Button
        variant="contained"
        className={classes.addButton}
        onClick={() => openDialog()}
      >
        {t('analyzer.addToPortfolio')}
      </Button>
      <Dialog
        onClose={() => setOpen(false)}
        aria-labelledby="dialog-title"
        open={open}
        className={classes.Dialog}
      >
        {/* The header of the dialog window */}
        <MuiDialogTitle disableTypography className={classes.dialogHeader}>
          <span className={classes.span}>
            <Typography variant="h6" className={classes.text}>
              {t('analyzer.dialogHeader')}
            </Typography>
          </span>
        </MuiDialogTitle>
        {/* The body of the dialog window */}
        <MuiDialogContent dividers className={classes.DialogContent}>
          <List className={classes.List} component="ol">
            {portfolios.map((portfolio) => (
              <ListEntry
                key={portfolio.id}
                id={portfolio.id}
                name={portfolio.name}
                checked={checked}
                handleToggle={handleCheckToggle}
                changeAmount={changeAmount}
                pattern={pattern}
              />
            ))}
          </List>
          {error && (
            <Typography color="error">
              <b>{t(errorTitleKey(error))}</b>: {t(errorMessageKey(error))}
            </Typography>
          )}
        </MuiDialogContent>
        {/* The footer of the dialog window */}
        <MuiDialogActions className={classes.dialogFooter}>
          <Button autoFocus onClick={() => setOpen(false)} color="primary">
            {t('portfolio.details.discardChanges')}
          </Button>
          <ProgressButton
            onClick={handleSave}
            color="primary"
            loading={loading}
            disabled={
              !Object.values(checked)
                .map((stringAmount) => pattern.test(stringAmount))
                .every((v) => v)
            }
          >
            {t('portfolio.details.saveChanges')}
          </ProgressButton>
        </MuiDialogActions>
      </Dialog>
    </div>
  );
};

// exports
export default AddToPortfolioButton;
