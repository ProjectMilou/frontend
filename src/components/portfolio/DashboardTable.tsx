import React from 'react';
import { useTranslation } from 'react-i18next';
import {
  Button,
  lighten,
  ListItemText,
  makeStyles,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  TableSortLabel,
  Theme,
  Typography,
} from '@material-ui/core';
import classNames from 'classnames';
import {
  NonEmptyPortfolioOverview,
  PortfolioOverview,
} from '../../portfolio/APIClient';
import DashboardActions from './DashboardActions';
import EuroCurrency from '../shared/EuroCurrency';
import Performance from '../shared/Performance';

const useStyles = makeStyles((theme: Theme) => ({
  action: { display: 'inline-block' },
  row: {
    cursor: 'pointer',
  },
  rowHover: {
    backgroundColor: lighten(theme.palette.primary.light, 0.85),
  },
  positionCount: {
    fontSize: '24px',
  },
  disabled: {
    cursor: 'not-allowed',
  },
  createButton: {
    marginTop: '25px',
  },
}));

type DashboardTableRowProps = {
  portfolio: PortfolioOverview;
  selectPortfolio: (id: string) => void;
  renamePortfolio: (id: string) => void;
  duplicatePortfolio: (id: string) => void;
  deletePortfolio: (id: string) => void;
};

const DashboardTableRow: React.FC<DashboardTableRowProps> = ({
  portfolio,
  selectPortfolio,
  renamePortfolio,
  duplicatePortfolio,
  deletePortfolio,
}) => {
  const [hover, setHover] = React.useState<boolean>(false);

  const { t } = useTranslation();
  const classes = useStyles();

  return (
    <TableRow
      onClick={() => {
        selectPortfolio(portfolio.id);
      }}
      onMouseEnter={() => setHover(true)}
      onMouseLeave={() => setHover(false)}
      className={classNames(classes.row, hover && classes.rowHover)}
    >
      <TableCell align="center">
        {'score' in portfolio && portfolio.score}
      </TableCell>
      <TableCell align="center" component="th" scope="row">
        <ListItemText
          primary={<Typography color="primary">{portfolio.name}</Typography>}
          secondary={
            portfolio.virtual ? t('portfolio.virtual') : t('portfolio.real')
          }
          secondaryTypographyProps={{
            color: portfolio.virtual ? 'textSecondary' : 'secondary',
          }}
        />
      </TableCell>
      <TableCell align="center" className={classes.positionCount}>
        {portfolio.positionCount}
      </TableCell>
      <TableCell align="center">
        <EuroCurrency value={portfolio.value} />
      </TableCell>
      <TableCell align="center">
        <Performance value={portfolio.perf7d} />
      </TableCell>
      <TableCell align="center">
        <Performance value={portfolio.perf1y} />
      </TableCell>
      <TableCell align="center">
        <DashboardActions
          portfolio={portfolio}
          renamePortfolio={renamePortfolio}
          duplicatePortfolio={duplicatePortfolio}
          deletePortfolio={deletePortfolio}
        />
      </TableCell>
    </TableRow>
  );
};

type Order = 'asc' | 'desc';

type KeysOfUnion<T> = T extends T ? keyof T : never;
type PortfolioOverviewKeys = KeysOfUnion<PortfolioOverview>;

interface SortableHeadCell {
  id: PortfolioOverviewKeys;
  label: string;
  numeric: boolean;
}

export type DashboardTableHeadProps = {
  onRequestSort: (property: PortfolioOverviewKeys) => void;
  order: Order;
  orderBy: PortfolioOverviewKeys;
};

export const DashboardTableHead: React.FC<DashboardTableHeadProps> = ({
  onRequestSort,
  order,
  orderBy,
}) => {
  const { t } = useTranslation();
  const headCells: SortableHeadCell[] = [
    { id: 'score', numeric: true, label: t('portfolio.score') },
    { id: 'name', numeric: false, label: t('portfolio.name') },
    {
      id: 'positionCount',
      numeric: true,
      label: t('portfolio.positionsCount'),
    },
    { id: 'value', numeric: true, label: t('portfolio.value') },
    { id: 'perf7d', numeric: true, label: t('portfolio.7d') },
    { id: 'perf1y', numeric: true, label: t('portfolio.1y') },
  ];

  return (
    <TableHead>
      <TableRow>
        {headCells.map((headCell) => (
          <TableCell
            align="center"
            key={headCell.id}
            sortDirection={orderBy === headCell.id ? order : false}
          >
            <TableSortLabel
              active={orderBy === headCell.id}
              direction={orderBy === headCell.id ? order : 'asc'}
              onClick={() => onRequestSort(headCell.id)}
            >
              {headCell.label}
            </TableSortLabel>
          </TableCell>
        ))}
        <TableCell align="center">{t('portfolio.actions')}</TableCell>
      </TableRow>
    </TableHead>
  );
};

export type DashboardTableProps = {
  portfolios: PortfolioOverview[];
  selectPortfolio: (id: string) => void;
  renamePortfolio: (id: string) => void;
  duplicatePortfolio: (id: string) => void;
  deletePortfolio: (id: string) => void;
  createPortfolio: () => void;
};

function comparatorTable<T>(a: T, b: T) {
  if (a < b) {
    return -1;
  }
  if (a > b) {
    return 1;
  }
  return 0;
}

function isDefined(obj: Record<string, unknown>, key: string) {
  return key in obj && obj[key] !== undefined;
}

function sortPortfolios(
  portfolios: PortfolioOverview[],
  orderBy: PortfolioOverviewKeys,
  order: Order
) {
  // portfolios without undefined values in the sorting column
  const nonUndefined = portfolios.filter((values) =>
    isDefined(values, orderBy)
  ) as NonEmptyPortfolioOverview[];

  const sorted =
    order === 'asc'
      ? nonUndefined.sort((a, b) => comparatorTable(a[orderBy], b[orderBy]))
      : nonUndefined.sort((a, b) => comparatorTable(b[orderBy], a[orderBy]));

  // undefined values are always at the end
  return [
    ...sorted,
    ...portfolios.filter((values) => !isDefined(values, orderBy)),
  ];
}

const DashboardTable: React.FC<DashboardTableProps> = ({
  portfolios,
  selectPortfolio,
  renamePortfolio,
  duplicatePortfolio,
  deletePortfolio,
  createPortfolio,
}) => {
  const [order, setOrder] = React.useState<Order>('desc');
  const [orderBy, setOrderBy] = React.useState<PortfolioOverviewKeys>('score');
  const [sortedPortfolios, setSortedPortfolios] = React.useState(() =>
    sortPortfolios(portfolios, orderBy, order)
  );

  // only sort portfolios when needed
  React.useEffect(() => {
    setSortedPortfolios(sortPortfolios(portfolios, orderBy, order));
  }, [portfolios, orderBy, order]);

  const classes = useStyles();
  const { t } = useTranslation();

  const handleRequestSort = (property: PortfolioOverviewKeys) => {
    const isAsc = orderBy === property && order === 'asc';
    setOrder(isAsc ? 'desc' : 'asc');
    setOrderBy(property);
  };
  // TODO: Improve portfolio score visualization
  return (
    <>
      {!!portfolios.length && (
        <TableContainer component={Paper}>
          <Table>
            <DashboardTableHead
              onRequestSort={handleRequestSort}
              order={order}
              orderBy={orderBy}
            />
            <TableBody>
              {sortedPortfolios.map((p) => (
                <DashboardTableRow
                  portfolio={p}
                  selectPortfolio={selectPortfolio}
                  key={p.id}
                  renamePortfolio={renamePortfolio}
                  duplicatePortfolio={duplicatePortfolio}
                  deletePortfolio={deletePortfolio}
                />
              ))}
            </TableBody>
          </Table>
        </TableContainer>
      )}
      <Button
        className={classes.createButton}
        variant="outlined"
        color="primary"
        onClick={() => createPortfolio()}
      >
        {t('portfolio.dashboard.createPortfolio')}
      </Button>
    </>
  );
};

export default DashboardTable;
