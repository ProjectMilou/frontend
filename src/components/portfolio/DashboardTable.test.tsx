import { render, fireEvent } from '@testing-library/react';
import * as React from 'react';
import DashboardTable, { DashboardTableProps } from './DashboardTable';
import { MockOverview, MockOverviewTwo } from '../../portfolio/APIMocks';
import { portfolioDetails } from '../../portfolio/Router';

describe('DashboardTable', () => {
  const defaultProps: DashboardTableProps = {
    portfolios: [MockOverview, MockOverviewTwo],
    renamePortfolio: jest.fn(),
    duplicatePortfolio: jest.fn(),
    deletePortfolio: jest.fn(),
    createPortfolio: jest.fn(),
  };

  const renderComponent = (newProps?: Partial<DashboardTableProps>) => {
    const props = { ...defaultProps, ...newProps };
    return {
      ...render(<DashboardTable {...props} />),
      props,
    };
  };

  test('show no table if no portfolios exist', () => {
    const { container } = renderComponent({ portfolios: [] });
    // no table
    expect(container.querySelector('table')).toBeNull();
  });

  test('display portfolio information', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('select portfolio on click', () => {
    const { props, getByText } = renderComponent();
    fireEvent.click(getByText(props.portfolios[0].name));
    expect(portfolioDetails).toHaveBeenCalledWith(props.portfolios[0].id);
    fireEvent.click(getByText(props.portfolios[1].name));
    expect(portfolioDetails).toHaveBeenCalledWith(props.portfolios[1].id);
  });

  test('rename portfolio on click', () => {
    const { props } = renderComponent();
    fireEvent.click(document.querySelectorAll('button svg')[0]);
    expect(props.renamePortfolio).toHaveBeenCalledWith(props.portfolios[0].id);
    fireEvent.click(document.querySelectorAll('button svg')[3]);
    expect(props.renamePortfolio).toHaveBeenCalledWith(props.portfolios[1].id);
    expect(portfolioDetails).not.toHaveBeenCalled();
  });

  test('duplicate portfolio on click', () => {
    const { props } = renderComponent();
    fireEvent.click(document.querySelectorAll('button svg')[1]);
    expect(props.duplicatePortfolio).toHaveBeenCalledWith(
      props.portfolios[0].id
    );
    fireEvent.click(document.querySelectorAll('button svg')[4]);
    expect(props.duplicatePortfolio).toHaveBeenCalledWith(
      props.portfolios[1].id
    );
    expect(portfolioDetails).not.toHaveBeenCalled();
  });

  test('delete portfolio on click', () => {
    const { props } = renderComponent();
    fireEvent.click(document.querySelectorAll('button svg')[5]);
    expect(props.deletePortfolio).not.toHaveBeenCalled();
    fireEvent.click(document.querySelectorAll('button svg')[2]);
    expect(props.deletePortfolio).toHaveBeenCalledWith(props.portfolios[0].id);
    expect(portfolioDetails).not.toHaveBeenCalled();
  });
});
