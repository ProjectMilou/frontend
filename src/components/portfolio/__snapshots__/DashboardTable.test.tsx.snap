// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardTable display portfolio information 1`] = `
<div>
  <div
    class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      class="MuiTable-root"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
        >
          <th
            aria-sort="descending"
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-active"
              role="button"
              tabindex="0"
            >
              portfolio.score
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionDesc"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiTableSortLabel-root"
              role="button"
              tabindex="0"
            >
              portfolio.name
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiTableSortLabel-root"
              role="button"
              tabindex="0"
            >
              portfolio.positionsCount
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiTableSortLabel-root"
              role="button"
              tabindex="0"
            >
              portfolio.value
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiTableSortLabel-root"
              role="button"
              tabindex="0"
            >
              portfolio.7d
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiTableSortLabel-root"
              role="button"
              tabindex="0"
            >
              portfolio.1y
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            portfolio.actions
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root makeStyles-row-8"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <div
              class="makeStyles-box-13 makeStyles-darkRed-17"
            >
              <p
                class="MuiTypography-root makeStyles-title-14 MuiTypography-body1"
              >
                1
              </p>
            </div>
          </td>
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            role="cell"
            scope="row"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary"
              >
                test
              </p>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                portfolio.virtual
              </p>
            </div>
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-positionCount-10 MuiTableCell-alignCenter"
          >
            3
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-styledNumberFormat-22 makeStyles-styledNumberFormat-23"
            >
              9,00 €
            </span>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-styledNumberFormat-22 makeStyles-styledNumberFormat-24"
            >
              -1,23 %
            </span>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-styledNumberFormat-22 makeStyles-styledNumberFormat-25"
            >
              13,37 %
            </span>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <div
              class="makeStyles-action-26"
              title="portfolio.rename"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              class="makeStyles-action-26"
              title="portfolio.duplicate"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M4 6H2v14c0 1.1.9 2 2 2h14v-2H4V6zm16-4H8c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-1 9h-4v4h-2v-4H9V9h4V5h2v4h4v2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              class="makeStyles-action-26"
              title="portfolio.delete"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root makeStyles-row-8"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <div
              class="makeStyles-box-13 makeStyles-darkRed-17"
            >
              <p
                class="MuiTypography-root makeStyles-title-14 MuiTypography-body1"
              >
                0.6
              </p>
            </div>
          </td>
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            role="cell"
            scope="row"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary"
              >
                testTwo
              </p>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorSecondary MuiTypography-displayBlock"
              >
                portfolio.real
              </p>
            </div>
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-positionCount-10 MuiTableCell-alignCenter"
          >
            4
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-styledNumberFormat-22 makeStyles-styledNumberFormat-28"
            >
              174,98 €
            </span>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-styledNumberFormat-22 makeStyles-styledNumberFormat-29"
            >
              0,00 %
            </span>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-styledNumberFormat-22 makeStyles-styledNumberFormat-30"
            >
              -1,00 %
            </span>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <div
              class="makeStyles-action-26"
              title="portfolio.rename"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              class="makeStyles-action-26"
              title="portfolio.duplicate"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M4 6H2v14c0 1.1.9 2 2 2h14v-2H4V6zm16-4H8c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-1 9h-4v4h-2v-4H9V9h4V5h2v4h4v2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              class="makeStyles-action-26 makeStyles-disabled-27"
              title="portfolio.deleteReal"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="makeStyles-buttons-12"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        portfolio.dashboard.createPortfolio
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        portfolio.dashboard.importPortfolio
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
